### ISSUE TITLE
Unexpected union type between generic and non-generic function

### ISSUE THREAD

**Issue by gcnew** [Microsoft/TypeScript#12970]
**TypeScript Version:**  nightly (2.2.0)

**Code**

```ts
function id<T>(x: T): T {
    return x;
}

function stringId(x: string): string {
    return x;
}

function numberId(x: number): number {
    return x;
}

// OK: ((x: string) => string) | ((x: number) => number)
const f = true ? numberId : stringId;

// Actual: <T>(x: T) => T
// Expected: (x: string) => string
const g = true ? id : stringId;
```

**Expected behavior:**
Union between a function and a generic function should yield the most specific type, not the most general one.

**Comment by JsonFreeman** [comment]
I agree with the expected behavior. I also wonder if it would be reasonable to not collapse the union and keep both constituents.

**Comment by gcnew** [comment]
Fixed by #16368
